### DISCUSSION TITLE
Computing Partial derivatives of a function

### DISCUSSION THREAD

**Discussion by Betancourt20** [stack-of-tasks/pinocchio#1894] [Q&A] [answered]
Hello, 

I was looking to the best/optimal way to compute ${ \frac{\partial{({J} {J}^T)}}{\partial{{q}_i}} }$ and I found that using finite difference is quite not the best as explained in https://hal.science/hal-01790971 . Do you believe there is another easiest way? 

I found this example https://github.com/stack-of-tasks/pinocchio/blob/master/unittest/python/bindings_com_velocity_derivatives.py but I'm not sure this is the best way to do it.

Thank you so much for your support,
best regards,
Julio

**Comment by jcarpent** [comment, marked as answer]
${ \frac{\partial{({J} {J}^T)}}{\partial{{q}_i}} }$ is a tensor. There is no real reason to store it from an efficiency perspective. Could you detail the scope where you would like to exploit such computation?

**Reply by Betancourt20** [reply]
Thanks for your answer. Indeed, I agree with you.
I'm trying to get some conditions to guarantee the stability of the system when a variable virtual mass is imposed. Thus, after some computations, I got the following condition
$$\frac{\partial{(J J^T)}}{\partial{q_i}} (J J^T)^{-1} > \frac{\partial{\kappa (q)}}{\partial{q_i}} I \frac{1}{\kappa(q)}$$
I want to do a use case where I can clarify with an ellipsoid, for instance, that in certain cases/configurations, that conditions does not hold if the $\kappa$ is not well defined. 
Thanks in advance,
Julio

**Reply by jcarpent** [reply]
For critical applications, we may think to compute these derivatives analytically of course. This will require some additional work.
I would suggest first obtaining them via finite differences at first glance. Using central FD will give you already sufficient precision to validate your pipeline.

**Reply by Betancourt20** [reply]
Thank you so much for your answer. Yes, I think is the best to do for the moment. 
Best regards,
Julio
